### PR TITLE
#163153952 Feature/163153952/login user

### PR DESCRIPTION
**Pull request by habibaudu**
#### What does this PR do?
Enable User to Login

#### Description of Task to be completed?
- Write controller to login user
- write unit Test for failures
- write Unit tests for success
- Ensure User can log in using either Email or Username
- Ensure user can log in with either uppercase or lower case email, username, password
- write validation for invalid credentials

#### How should this be manually tested?
- Fire Up postman 
- URL: http://localhost:3000/api/v1/users/login
- Enter username or email and password.
- click send

#### What are the relevant pivotal tracker stories?
- [#163153952](https://www.pivotaltracker.com/story/show/163153952)

### CODE REVIEW

File: controllers/user.js [hunk 1]
```diff
@@ -66,31 +66,43 @@ class Users {
...
-  static login(req, res, next) {
-    passport.authenticate('local', { session: false }, (
-      err,
-      user
-    ) => {
-      if (err) {
-        return next(err);
-      }
+  static async login(req, res) {
+    const usernameOrEmail = (req.body.email)
```
Comment (oreoluwade) [inline controllers/user.js]: The logic here is not quite accurate. Yes, you should get two properties from `req.body` which may be (as you have tried to) ``usernameOrEmail` and then `password`. 

At the point of finding the user with `User.findOne`, what you'd be doing is finding the user when `username` is equal to `req.body.usernameOrEmail` or email is equal to `req.body.usernameOrEmail`.

You'd have to use sequelize operators to accomplish this.

@rajeman already used one sequelize operator in the user validator, so he might be able to provide some context. You'd use two in your case; the `or` operator and the `iLike` operator.

Check sequelize documentation for the use of operators. Also, pair with @rajeman to implement this.

File: controllers/user.js [hunk 2]
```diff
@@ -66,31 +66,43 @@ class Users {
...
+      ? { email: req.body.email.toLowerCase() }
+      : { username: req.body.username.toLowerCase() };
+    const loginUser = await User
+      .findOne({ where: usernameOrEmail });
+    if (loginUser) {
+      if (bcrypt.compareSync(req.body.password.toLowerCase(),
```
Comment (oreoluwade) [inline controllers/user.js]: The password should not be mutated with any string method; `toLowerCase`, `toUpperCase` or to anything at all. It should remain as it was supplied, else there won't be a match when comparing passwords.

File: controllers/user.js [hunk 3]
```diff
@@ -66,31 +66,43 @@ class Users {
...
     * @param {object} req - The request object.
     * @param {object} res - The response object.
     */
   static async register(req, res) {
-    const { username, email, password } = req.body;
-
+    let { username, email, password } = req.body;
+    username = username.toLowerCase();
+    email = email.toLowerCase();
+    password = password.toLowerCase();
```
Comment (oreoluwade) [inline controllers/user.js]: Same thing here. Don't mutate the password at the point of registration.

File: middlewares/userValidations.js [hunk 4]
```diff
@@ -109,6 +109,46 @@ export default {
       return next(error);
     }
     next();
+  },
+  loginParamsValidator(req, res, next) {
+    const availableParams = (req.body.email)
```
Comment (oreoluwade) [inline middlewares/userValidations.js]: The params you should be checking for while logging in will be `usernameOrEmail` and `password` to rhyme with what is in the controller.

File: controllers/user.js (hunk 1 above)
Comment (habibaudu) [reply controllers/user.js]: ok... will do

File: middlewares/userValidations.js (hunk 4 above)
Comment (habibaudu) [reply middlewares/userValidations.js]: okay... thanks

File: controllers/user.js (hunk 2 above)
Comment (habibaudu) [reply controllers/user.js]: ok thanks, will change it.

File: controllers/user.js [hunk 5]
```diff
@@ -66,22 +69,28 @@ class Users {
...
-      }
-
-      if (user) {
-        return res.json({ message: 'Login was successful' });
+  static async login(req, res) {
+    const loginUser = await User
+      // eslint-disable-next-line max-len
+      .findOne({ where: { [or]: [{ username: { [iLike]: req.body.username } }, { email: { [iLike]: req.body.email } }] } });
+    if (loginUser) {
+      if (bcrypt.compareSync(req.body.password, loginUser.password)) {
```
Comment (oaboluwarin) [inline controllers/user.js]: There is a `validPassword` method available on the Prototype of the `User` model. The function encapsulates the password checking functionality. That's what you should use to check the password here. No need to import `bcrypt`.

TLDR: Check out the `validPassword` method in the `User` model for more context.

Comment (habibaudu) [reply controllers/user.js]: ok.thanks.

File: controllers/user.js [hunk 6]
```diff
@@ -66,22 +69,22 @@ class Users {
...
+    const loginUser = await User
+      .findOne({
+        where: {
+          [or]: [{ username: { [iLike]: usernameOrEmail } },
+            { email: { [iLike]: usernameOrEmail } }]
+        }
+      });
+    await loginUser.validPassword(password);
+    return res.status(200).json({
+      message: 'Login was sucessful'
```
Comment (oreoluwade) [inline controllers/user.js]: Please correct the spelling of successful here.

File: controllers/user.js [hunk 7]
```diff
@@ -66,22 +69,22 @@ class Users {
...
-      if (user) {
-        return res.json({ message: 'Login was successful' });
-      }
-      return res.status(401).json('incorrect email or password');
+  static async login(req, res) {
+    const { usernameOrEmail, password } = req.body;
+    const loginUser = await User
+      .findOne({
+        where: {
+          [or]: [{ username: { [iLike]: usernameOrEmail } },
```
Comment (oreoluwade) [inline controllers/user.js]: Step the `username` object to the next line with its curly braces so that it aligns with the email object.

File: controllers/user.js [hunk 8]
```diff
@@ -66,22 +69,22 @@ class Users {
...
+            { email: { [iLike]: usernameOrEmail } }]
+        }
+      });
+    await loginUser.validPassword(password);
+    return res.status(200).json({
+      message: 'Login was sucessful'
     });
   }
 
+
```
Comment (oreoluwade) [inline controllers/user.js]: Remove this extra line. One line is enough between two methods.

File: middlewares/userValidations.js [hunk 9]
```diff
@@ -109,6 +109,69 @@ export default {
...
+    if (!errorArray.length) {
+      return next();
+    }
+
+    const errorMessage = JSON.stringify(errorArray);
+    const error = new Error(errorMessage);
+    error.status = 400;
+    return next(error);
+  },
+  async invalidCredential(req, res, next) {
```
Comment (oreoluwade) [inline middlewares/userValidations.js]: Please pluralize this `invalidCredential` middleware

File: middlewares/userValidations.js [hunk 10]
```diff
@@ -109,6 +109,69 @@ export default {
...
+    const { usernameOrEmail, password } = req.body;
+    const loginUser = await User
+      .findOne({
+        where: {
+          [or]: [{ username: { [iLike]: usernameOrEmail } },
```
Comment (oreoluwade) [inline middlewares/userValidations.js]: Step the username object down like before.

File: middlewares/userValidations.js [hunk 11]
```diff
@@ -109,6 +109,69 @@ export default {
...
+            { email: { [iLike]: usernameOrEmail } }]
+        }
+      });
+
+    if (!loginUser) {
+      const error = new Error('Invalid Credential');
```
Comment (oreoluwade) [inline middlewares/userValidations.js]: Pluralize `credential`

File: middlewares/userValidations.js [hunk 12]
```diff
@@ -109,6 +109,69 @@ export default {
...
+      error.status = 400;
+      return next(error);
+    }
+    const valid = await loginUser.validPassword(password);
```
Comment (oreoluwade) [inline middlewares/userValidations.js]: Since the validPassword method on the user object returns a boolean, it may be best to make the variable name more inclined towards that. For example, `passwordIsValid`

**Review by oreoluwade — CHANGES_REQUESTED** [review]

**Review by oreoluwade — APPROVED** [review]
LGTM
